webapp/src/components/SingleStudy/MapView/index.tsx

```diff
@@ -0,0 +1,391 @@
+  }
+  if (textSize <= 6) {
+    return FONT_SIZE * textSize * 8.5;
+  }
+  if (textSize <= 10) {
+    return FONT_SIZE * textSize * 7.5;
+  }
+  return FONT_SIZE * textSize * 6.5;
+};
+
+const GraphViewMemo = React.memo(GraphView);
+
+const MapView = (props: Props) => {
+  const classes = useStyles();
+  const [t] = useTranslation();
+  const { study } = props;
+  const [loaded, setLoaded] = useState(false);
+  const [selectedItem, setSelectedItem] = useState<NodeProperties | LinkProperties>();
+  const [selectedNodeLinks, setSelectedNodeLinks] = useState<Array<LinkProperties>>();
+  const [nodeData, setNodeData] = useState<Array<NodeProperties>>([]);
+  const [linkData, setLinkData] = useState<Array<LinkProperties>>([]);
+  const { enqueueSnackbar } = useSnackbar();
+  const [openModal, setOpenModal] = useState<boolean>(false);
+  const [firstNode, setFirstNode] = useState<string>();
+  const [secondNode, setSecondNode] = useState<string>();
+  const graphRef = useRef<Graph<GraphNode & NodeProperties, GraphLink & LinkProperties>>(null);
+  const prevselectedItemId = useRef<string>();
+  // const [zoom, setZoom] = useState<number>();
+
+  const onClickNode = useCallback((nodeId: string) => {
+    if (!firstNode && nodeData) {
+      const obj = nodeData.find((o) => o.id === nodeId);
+      setSelectedItem(obj);
+    } else if (firstNode) {
+      setSecondNode(nodeId);
+    }
+  }, [firstNode, nodeData]);
+
+  const onClickLink = useCallback((source: string, target: string) => {
+    const obj = {
+      source,
+      target,
+    };
+    setSelectedItem(obj);
+  }, []);
+
+  const createModeLink = useCallback((id: string) => {
+    if (firstNode && firstNode === id) {
+      setFirstNode(undefined);
+      setSecondNode(undefined);
+    } else {
+      setFirstNode(id);
+    }
+  }, [firstNode, setFirstNode, setSecondNode]);
+
+  const onClose = () => {
+    setOpenModal(false);
+  };
+
+  const onSave = async (name: string, posX: number, posY: number, color: string) => {
+    setOpenModal(false);
+    try {
+      await createArea(study.id, name);
+      setNodeData([...nodeData, {
+        id: name,
+        x: posX,
+        y: posY,
+        color,
+        rgbColor: color.slice(4, -1).split(',').map(Number),
+        size: { width: calculateSize(name), height: NODE_HEIGHT },
+      }]);
+    } catch (e) {
+      setNodeData(nodeData.filter((o) => o.id !== name));
+      enqueueErrorSnackbar(enqueueSnackbar, t('singlestudy:createAreaError'), e as AxiosError);
+    }
+  };
+
+  const handleUpdate = async (id: string, value: UpdateAreaUi) => {
+    const filterNodes = nodeData.find((o) => o.id === id);
+    if (filterNodes) {
+      try {
+        const prevColors = filterNodes.rgbColor;
+        if (value.color_rgb[0] !== prevColors[0] || value.color_rgb[1] !== prevColors[1] || value.color_rgb[2] !== prevColors[2]) {
+          const updateNode = nodeData.filter((o) => o.id !== id);
+          setNodeData([...updateNode, {
+            id,
```

Comment:
you can do : "...targetNode" instead of "id" since there may be other properties we want to keep